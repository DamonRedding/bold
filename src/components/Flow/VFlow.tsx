import React from 'react'

import { Flow, FlowProps } from './Flow'

export interface VFlowProps extends FlowProps {}

export function VFlow(props: VFlowProps) {
  return <Flow direction='vertical' vSpacing={1} hSpacing={0} {...props} />
}
