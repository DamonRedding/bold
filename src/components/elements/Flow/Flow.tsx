import { Interpolation } from 'emotion'
import React, { CSSProperties } from 'react'

import { Theme, useStyles } from '../../../styles'
import { AlignItems, JustifyContent } from '../Grid/Grid'

export interface FlowProps {
  direction?: 'horizontal' | 'vertical'
  vSpacing?: number
  hSpacing?: number
  alignItems?: AlignItems
  justifyContent?: JustifyContent
  style?: Interpolation
  children?: React.ReactNode
}

export const Flow = (props: FlowProps) => {
  const { direction, style } = props
  const { classes, css } = useStyles(createStyles, props)

  const renderChild = (child: any) => {
    return child && <div className={classes.child}>{child}</div>
  }

  const className = css(classes.flow, direction === 'horizontal' ? classes.flowHorizontal : classes.flowVertical, style)

  return <div className={className}>{React.Children.map(props.children, child => renderChild(child))}</div>
}

Flow.defaultProps = {
  direction: 'horizontal',
  vSpacing: 0,
  hSpacing: 1,
} as Partial<FlowProps>

export const createStyles = (
  theme: Theme,
  { alignItems, justifyContent, direction, vSpacing, hSpacing }: FlowProps
) => ({
  flow: {
    display: 'flex',
    alignItems,
    justifyContent,
  } as CSSProperties,
  flowHorizontal: {} as CSSProperties,
  flowVertical: {
    flexDirection: 'column',
  } as CSSProperties,
  child: {
    marginBottom: `${vSpacing / 2}rem`,
    marginTop: `${vSpacing / 2}rem`,
    marginLeft: `${hSpacing / 2}rem`,
    marginRight: `${hSpacing / 2}rem`,
    display: 'inline-block',

    ':first-of-type': {
      marginTop: direction === 'vertical' && 0,
      marginLeft: direction === 'horizontal' && 0,
    },
    ':last-of-type': {
      marginBottom: direction === 'vertical' && 0,
      marginRight: direction === 'horizontal' && 0,
    },
    ':empty': {
      display: 'none',
    },
  } as CSSProperties,
})
