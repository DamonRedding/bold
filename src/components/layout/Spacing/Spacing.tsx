import { Interpolation } from 'emotion'
import React from 'react'

import { withStyles, WithStylesProps } from '../../../styles/index'

export interface SpacingProps extends WithStylesProps {
    top?: number
    right?: number
    bottom?: number
    left?: number
    block?: boolean
    style?: Interpolation
}

@withStyles
export class Spacing extends React.PureComponent<SpacingProps> {

    static defaultProps: SpacingProps = {
        block: false,
    }

    render() {
        const styles = {
            marginTop: this.props.top + 'rem',
            marginRight: this.props.right + 'rem',
            marginBottom: this.props.bottom + 'rem',
            marginLeft: this.props.left + 'rem',
            display: !this.props.block && 'inline-block',
        }

        return (
            <div className={this.props.css(styles, this.props.style)}>{this.props.children}</div>
        )
    }
}
