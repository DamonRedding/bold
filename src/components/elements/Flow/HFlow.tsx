import React from 'react'

import { Flow, FlowProps } from './Flow'

export interface HFlowProps extends FlowProps {}

export function HFlow(props: HFlowProps) {
  return <Flow direction='horizontal' vSpacing={0} hSpacing={1} {...props} />
}
