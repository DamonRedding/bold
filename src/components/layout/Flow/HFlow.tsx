import React from 'react'

import { withStyles } from '../../../styles'

import { Flow, FlowProps } from './Flow'

export interface HFlowProps extends FlowProps {
}

@withStyles
export class HFlow extends React.PureComponent<HFlowProps> {

    render() {
        return (
            <Flow direction='horizontal' vSpacing={0} hSpacing={1} {...this.props}>
                {this.props.children}
            </Flow>
        )
    }
}
