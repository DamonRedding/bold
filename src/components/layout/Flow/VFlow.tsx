import React from 'react'

import { withStyles } from '../../../styles'

import { Flow, FlowProps } from './Flow'

export interface VFlowProps extends FlowProps {
}

@withStyles
export class VFlow extends React.PureComponent<VFlowProps> {

    render() {
        return (
            <Flow direction='vertical' vSpacing={1} hSpacing={0} {...this.props}>
                {this.props.children}
            </Flow>
        )
    }
}
