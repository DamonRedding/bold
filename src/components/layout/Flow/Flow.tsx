import * as React from 'react'

import { withStyles, WithStylesProps } from '../../../styles'
import { Spacing } from '../Spacing/Spacing'

export interface FlowProps extends WithStylesProps {
    direction?: 'horizontal' | 'vertical'
    vSpacing?: number
    hSpacing?: number
}

@withStyles
export class Flow extends React.PureComponent<FlowProps> {

    static defaultProps: FlowProps = {
        direction: 'horizontal',
        vSpacing: 0,
        hSpacing: 1,
    }

    render() {
        const { css, direction, hSpacing } = this.props

        const styles = {
            flow: {
                margin: `0 ${-hSpacing / 2}rem`,
            },
            flowHorizontal: {
                display: 'flex',
                flexWrap: 'wrap',
            },
            flowVertical: {
                display: 'flex',
                flexDirection: 'column',
            },
        }

        return (
            <div
                className={css(styles.flow, direction === 'horizontal' ? styles.flowHorizontal : styles.flowVertical)}
            >
                {React.Children.map(this.props.children, (child) => this.renderChild(child))}
            </div>
        )
    }

    renderChild = (child) => {
        return (
            <Spacing
                top={this.props.vSpacing / 2}
                bottom={this.props.vSpacing / 2}
                left={this.props.hSpacing / 2}
                right={this.props.hSpacing / 2}
            >
                {child}
            </Spacing>
        )
    }
}
