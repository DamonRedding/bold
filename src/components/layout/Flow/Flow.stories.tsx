import { number, select } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { Button } from '../../elements/Button'
import { AlignItems, JustifyContent } from '../Grid/Grid'

import { Flow } from './Flow'
import { HFlow } from './HFlow'
import { VFlow } from './VFlow'

const directionOptions = {
    'horizontal': 'horizontal',
    'vertical': 'vertical',
}

const justifyContentOptions: { [key in JustifyContent]: string } = {
    'flex-start': 'flex-start',
    'flex-end': 'flex-end',
    'center': 'center',
    'space-between': 'space-between',
    'space-around': 'space-around',
    'space-evenly': 'space-evenly',
}

const alignItemsOptions: { [key in AlignItems]: string } = {
    'flex-start': 'flex-start',
    'flex-end': 'flex-end',
    'center': 'center',
    'baseline': 'baseline',
    'stretch': 'stretch',
}

storiesOf('Layout', module)
    .add('Flow', () => (
        <>
            <Flow
                hSpacing={number('hSpacing', 1)}
                vSpacing={number('vSpacing', 0)}
                direction={select('direction', directionOptions, 'horizontal')}
                alignItems={select('alignItems', alignItemsOptions, 'stretch')}
                justifyContent={select('justifyContent', justifyContentOptions, 'flex-start')}
            >
                <span>1</span>
                <span>2</span>
                <span>3</span>
                <span>4</span>
            </Flow>
            <hr />
            <Flow
                hSpacing={number('hSpacing', 1)}
                vSpacing={number('vSpacing', 0)}
                direction={select('direction', directionOptions, 'horizontal')}
                alignItems={select('alignItems', alignItemsOptions, 'stretch')}
                justifyContent={select('justifyContent', justifyContentOptions, 'flex-start')}
            >
                <Button label='Botão 1' type='primary' />
                <Button label='Botão 2' />
                <Button label='Botão 3' />
            </Flow>
        </>
    ))
    .add('VFlow', () => (
        <VFlow>
            <Button label='Botão 1' type='primary' />
            <Button label='Botão 2' />
            <Button label='Botão 3' />
        </VFlow>
    ))
    .add('HFlow', () => (
        <HFlow>
            <Button label='Botão 1' type='primary' />
            <Button label='Botão 2' />
            <Button label='Botão 3' />
        </HFlow>
    ))
