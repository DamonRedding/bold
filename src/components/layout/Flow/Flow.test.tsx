import { mount, render, shallow } from 'enzyme'
import * as React from 'react'

import { withTheme } from '../../../test'
import { Spacing } from '../Spacing/Spacing'

import { Flow } from './Flow'

describe('Flow', () => {
    it('should wrap each child on an element with horizontal spacing', () => {
        const wrapper = render(withTheme(
            <Flow>
                <span>1</span>
                <span>2</span>
                <span>3</span>
            </Flow>
        ))
        expect(wrapper).toMatchSnapshot()
    })
    it('should accept vSpacing prop', () => {
        const wrapper = render(withTheme(
            <Flow vSpacing={2.5}>
                <span>1</span>
                <span>2</span>
                <span>3</span>
            </Flow>
        ))
        expect(wrapper).toMatchSnapshot()
    })
    it('should accept hSpacing prop', () => {
        const wrapper = render(withTheme(
            <Flow hSpacing={0.5}>
                <span>1</span>
                <span>2</span>
                <span>3</span>
            </Flow>
        ))
        expect(wrapper).toMatchSnapshot()
    })
    it('should accept direction prop', () => {
        const wrapper = render(withTheme(
            <Flow direction='vertical'>
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </Flow>
        ))
        expect(wrapper).toMatchSnapshot()
    })
    it('should accept justifyContent prop', () => {
        const wrapper = render(withTheme(
            <Flow justifyContent='flex-end'>
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </Flow>
        ))
        expect(wrapper).toMatchSnapshot()
    })
    it('should accept alignItems prop', () => {
        const wrapper = render(withTheme(
            <Flow alignItems='center'>
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </Flow>
        ))
        expect(wrapper).toMatchSnapshot()
    })
    it('should not wrap empty elements', () => {
        expect(mount(withTheme(
            <Flow>
                <div>1</div>
                {false && <div>2</div>}
                <div>3</div>
            </Flow>
        )).find(Spacing).length).toEqual(2)
    })
    it('should accept the style prop', () => {
        const wrapper = render(withTheme(
            <Flow style={{ padding: '10rem' }}>
                <div>1</div>
                <div>2</div>
            </Flow>
        ))
        expect(wrapper).toMatchSnapshot()
    })
})
