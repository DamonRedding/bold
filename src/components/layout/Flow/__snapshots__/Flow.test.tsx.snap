// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow deve envelopar cada filho em um elemento que crie espaçamento horizontal 1`] = `
<div
  class="css-1k5rx2j"
>
  <div
    class="css-1ueujd4"
  >
    <span>
      1
    </span>
  </div>
  <div
    class="css-1ueujd4"
  >
    <span>
      2
    </span>
  </div>
  <div
    class="css-1ueujd4"
  >
    <span>
      3
    </span>
  </div>
</div>
`;

exports[`Flow deve suportar o parâmetro direction 1`] = `
<div
  class="css-1n0xhs7"
>
  <div
    class="css-1ueujd4"
  >
    <div>
      1
    </div>
  </div>
  <div
    class="css-1ueujd4"
  >
    <div>
      2
    </div>
  </div>
  <div
    class="css-1ueujd4"
  >
    <div>
      3
    </div>
  </div>
</div>
`;

exports[`Flow deve suportar o parâmetro hSpacing 1`] = `
<div
  class="css-16nph7i"
>
  <div
    class="css-w6y3nz"
  >
    <span>
      1
    </span>
  </div>
  <div
    class="css-w6y3nz"
  >
    <span>
      2
    </span>
  </div>
  <div
    class="css-w6y3nz"
  >
    <span>
      3
    </span>
  </div>
</div>
`;

exports[`Flow deve suportar o parâmetro vSpacing 1`] = `
<div
  class="css-1k5rx2j"
>
  <div
    class="css-pmvu5p"
  >
    <span>
      1
    </span>
  </div>
  <div
    class="css-pmvu5p"
  >
    <span>
      2
    </span>
  </div>
  <div
    class="css-pmvu5p"
  >
    <span>
      3
    </span>
  </div>
</div>
`;
